File: editor/en/[725]Split Linked List in Parts.java

//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
    public ListNode[] splitListToParts(ListNode head, int k) {
        ListNode curr = head;
        int count = 0;

        while (curr != null) {
            curr = curr.next;
            count++;
        }

        ListNode[] result = new ListNode[k];
        int width = count / k;
        int reminder = count % k;
        curr = head;

        for (int i = 0; i < k; i++) {
            result[i] = curr;
            int currSize = width + ((i < reminder) ? 1 : 0);

            for (int j = 0; j < currSize - 1; j++) {
                if (curr != null) {
                    curr = curr.next;
                }
            }

            if (curr != null) {
                ListNode next = curr.next;
                curr.next = null;
                curr = next;
            }
        }

        return result;
    }
}
//leetcode submit region end(Prohibit modification and deletion)

File: editor/en/[725]Split Linked List in Parts.js
//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode[]}
 */
var splitListToParts = function (head, k) {
    let curr = head;
    let count = 0;
    while (curr) {
        curr = curr.next;
        count++;
    }
    let width = Math.floor(count / k);
    let reminder = count % k;
    curr = head;
    let list = new Array();
    for (let i = 0; i < k; i++) {
        let headNode = null;
        let tailNode = null;
        for (let j = 0; j < width + (i < reminder ? 1 : 0); j++) {
            if (headNode === null) {
                headNode = tailNode = curr;
            } else {
                tailNode.next = curr;
                tailNode = tailNode.next;
            }
            if (curr !== null) {
                curr = curr.next;
            }
        }
        if (tailNode !== null) {
            tailNode.next = null;
        }
        list[i] = headNode;
    }
    return list;
};
//leetcode submit region end(Prohibit modification and deletion)
